variables.tf

```diff
@@ -14,104 +14,120 @@
  * limitations under the License.
  */
 
-variable project {
+variable "project" {
   description = "The project to deploy to, if not set the default provider project is used."
-  default     = ""
+  type        = string
 }
 
-variable region {
+variable "region" {
   description = "Region for cloud resources"
+  type        = string
   default     = "us-central1"
 }
 
-variable ip_version {
+variable "ip_version" {
   description = "IP version for the Global address (IPv4 or v6) - Empty defaults to IPV4"
+  type        = string
   default     = ""
 }
 
-variable firewall_networks {
+variable "firewall_networks" {
   description = "Names of the networks to create firewall rules in"
-  type        = "list"
-  default     = ["default"]
+  type        = list(string)
+  default     = [
+    "default"]
 }
 
-variable firewall_projects {
+variable "firewall_projects" {
   description = "Names of the projects to create firewall rules in"
-  type        = "list"
-  default     = ["default"]
+  type        = list(string)
+  default     = [
+    "default"]
 }
 
-variable name {
+variable "name" {
   description = "Name for the forwarding rule and prefix for supporting resources"
+  type        = string
 }
 
-variable target_tags {
+variable "target_tags" {
   description = "List of target tags for health check firewall rule."
-  type        = "list"
+  type        = list(string)
 }
 
-variable backends {
+variable "backends" {
   description = "Map backend indices to list of backend maps."
-  type        = "map"
+  type        = map(list(object({
+    group = sting
```

Comment:
@aaron-lane, did you want to make them required? Otherwise we should probably leave previous version with map of strings..

The full validation doesn't work:
![image](https://user-images.githubusercontent.com/50832464/65312346-f9913380-db9a-11e9-9c02-438c2d234e46.png)

The validation with only `group` required does:
![image](https://user-images.githubusercontent.com/50832464/65312642-9d7adf00-db9b-11e9-8c2f-468ac1020c7a.png)

The rest of attributes [are not listed as required or listed under specific conditions](https://cloud.google.com/compute/docs/reference/rest/v1/backendServices/insert)

Reply:
Let's use an object with all supported keys. There is a improvement in the works to support optional object keys with default values, but any undesired key can be passed a value of `null` at present. Using a map is misleading because it implies that any key could be valid which is not true.

Reply:
@aaron-lane but what if they're not undesired? Dynamic block supports _all_ possible values. If we null them in variables users will have to guess that a var needs to be enabled in the variables file.
Should we document that in readme and/or leave a comment with guidance in the code?

Reply:
@aaron-lane the `null` doesn't work as of now (terraform 0.12.9)
![image](https://user-images.githubusercontent.com/50832464/65773734-7e54f200-e145-11e9-814e-361ea45487ae.png)
If we we make them required with string type that would brake all the current examples:
as the issue I described [here](https://github.com/terraform-google-modules/terraform-google-lb-http/pull/52#discussion_r326528997) with full validation is still valid.

How do you feel about creating separated issue which would take the awaited [terraform feature](https://github.com/hashicorp/terraform/issues/19898) as a dependency ?